src/net_processing.cpp

```diff
@@ -4693,33 +4693,33 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
 
         // Detect whether we're stalling
         current_time = GetTime<std::chrono::microseconds>();
-        if (state.nStallingSince && state.nStallingSince < count_microseconds(current_time) - 1000000 * BLOCK_STALLING_TIMEOUT) {
+        if (state.m_stalling_since.count() && state.m_stalling_since < current_time - BLOCK_STALLING_TIMEOUT) {
             // Stalling only triggers when the block download window cannot move. During normal steady state,
             // the download window should be much larger than the to-be-downloaded set of blocks, so disconnection
             // should only happen during initial block download.
             LogPrintf("Peer=%d is stalling block download, disconnecting\n", pto->GetId());
             pto->fDisconnect = true;
             return true;
         }
-        // In case there is a block that has been in flight from this peer for 2 + 0.5 * N times the block interval
+        // In case there is a block that has been in flight from this peer for block_interval * (1 + 0.5 * N)
         // (with N the number of peers from which we're downloading validated blocks), disconnect due to timeout.
         // We compensate for other peers to prevent killing off peers due to our own downstream link
         // being saturated. We only count validated in-flight blocks so peers can't advertise non-existing block hashes
         // to unreasonably increase our timeout.
         if (state.vBlocksInFlight.size() > 0) {
             QueuedBlock &queuedBlock = state.vBlocksInFlight.front();
             int nOtherPeersWithValidatedDownloads = nPeersWithValidatedDownloads - (state.nBlocksInFlightValidHeaders > 0);
-            if (count_microseconds(current_time) > state.nDownloadingSince + consensusParams.nPowTargetSpacing * (BLOCK_DOWNLOAD_TIMEOUT_BASE + BLOCK_DOWNLOAD_TIMEOUT_PER_PEER * nOtherPeersWithValidatedDownloads)) {
+            if (current_time > state.m_downloading_since + std::chrono::seconds{consensusParams.nPowTargetSpacing} * (BLOCK_DOWNLOAD_TIMEOUT_BASE + BLOCK_DOWNLOAD_TIMEOUT_PER_PEER * nOtherPeersWithValidatedDownloads)) {
```

Comment:
> which doesn't really seem right to me.

Do you mean mathematically or conceptually on a meta level?

Mathematically, it seems correct. Since all variables are constants, we have:

* Before `600 * ( 1000000 + 500000 * nOtherPeersWithValidatedDownloads)` (unit is `us`, implicitly)
* After `600 * (1. + .5 * nOtherPeersWithValidatedDownloads)`, (unit is `s`, explicitly. Converted by the compiler to `us` by multiplying with 1kk)

The difference between the two is zero:

https://www.wolframalpha.com/input/?i=%28+600+*+%28+1000000+%2B+500000+*+x+%29+%29+-+%28+1000000+*+%28+600+*+%28+1+%2B+0.5+*+x+%29+%29+%29

Reply:
As @MarcoFalke mentioned, the units do make sense to me. Also BLOCK_DOWNLOAD_TIMEOUT_BASE in multiples of block intervals seems easier to understand?

Reply:
"doesn't really seem right" -- yeah, more along the lines of "morally right" vs "mathematically correct" :)

In general using doubles where they're not needed just seems like a bad idea to me -- there's way more possible special cases and pitfalls with them, and the code already works fine without them, so why change it? Changing behaviour in potentially subtle ways in refactors bothers me in general -- to me that seems like the easiest way for a potential attacker to slip vulnerabilities into the codebase.

For example, in this line, I think behind the scenes chrono is:

 * converting the `chrono::seconds` version of `nPowTargetSpacing` to `duration<double,ratio=1,1>` when multiplying by a double
 * converting both `m_downloading_since` and the result of the above to `duration<double,ratio=1,1000000>` (converting one to a double, multiplying the other by 1M)
 * converting `current_time` to the `duration<double,ratio=1,1000000>` (ie conveting to a double) as well and then doing a floating point comparison between the two

Are those all fine? Yeah, I think so -- but why add the burden of reviewing all that just to use "0.5" instead of "500000"? In general, figuring out what the code actually does is the tricky part, so  making it easy to see what the plan was and hard to see what the compiler's going to do seems backwards to me in general. After all, if it's not clear from the code what the plan was, you can add a comment.

Reply:
What I'd prefer is something more like:

```c++
constexpr Fraction BLOCK_DOWNLOAD_TIMEOUT_BASE{1};
constexpr Fraction BLOCK_DOWNLOAD_TIMEOUT_PER_PEER{1,2};
```

so that you can define the constants meaningfully, but still have all the calculations done with integer math. You can make that actually work relatively easily:

```c++
class Fraction {
public:
    constexpr Fraction(int64_t n, int64_t d) : num{n}, den{d} { }
    explicit constexpr Fraction(int64_t n) : num{n}, den{1} { }
    int64_t num = 0, den = 1;
};

constexpr Fraction operator+(Fraction l, Fraction r) { return Fraction{l.num*r.den + r.num*l.den, l.den*r.den}; }
constexpr Fraction operator*(Fraction l, int64_t n) { return Fraction{l.num*n, l.den}; }
constexpr std::chrono::microseconds operator*(std::chrono::microseconds t, Fraction f) { return t*f.num/f.den; }
```

but maintaining a special class for this seems crazy, and C++ stdlib seems to only provide compile-time rationals (via std::ratio) which is useless here, and boost::rational seems to insist on normalising the denominator for no particular reason (and the fixedpoint stuff seems even more overkill). Still not convinced having floating point math is a win compared to just having 1M and 500k as constants, but there doesn't seem to be any better alternatives than those.